src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,407 @@
+    if ((id != "all") && !searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    QJsonArray statusArray;
+
+    if (id == "all") {
+        for (const QString &searchId : searchHandlers.keys()) {
+            const SearchHandlerPtr searchHandler = searchHandlers[searchId];
+            statusArray << QJsonObject {
+                {"id", searchId},
+                {"status", searchHandler->isActive() ? "Running" : "Stopped"},
+                {"total", searchHandler->results().size()}
+            };
+        }
+    }
+    else {
+        const SearchHandlerPtr searchHandler = searchHandlers[id];
+        statusArray << QJsonObject {
+            {"id", id},
+            {"status", searchHandler->isActive() ? "Running" : "Stopped"},
+            {"total", searchHandler->results().size()}
+        };
+    }
+
+    setResult(statusArray);
+}
+
+void SearchController::resultsAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    int limit = params()["limit"].toInt();
+    int offset = params()["offset"].toInt();
+
+    auto searchHandlers = sessionManager()->session()->getData(SEARCH_HANDLERS).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+    const QList<SearchResult> searchResults = searchHandler->results();
+    const int size = searchResults.size();
+
+    if ((offset > size) || ((offset == size) && (offset != 0)))
+        throw APIError(APIErrorType::Conflict, tr("Offset is out of range"));
+
+    // normalize values
+    if (offset < 0)
+        offset = size + offset;
+    if (offset < 0)  // check again
+        throw APIError(APIErrorType::Conflict);
+    if (limit <= 0)
+        limit = -1;
+
+    if ((limit > 0) || (offset > 0))
+        setResult(getResults(searchResults.mid(offset, limit), searchHandler->isActive(), size));
+    else
+        setResult(getResults(searchResults, searchHandler->isActive(), size));
+}
+
+void SearchController::deleteAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    ISession *session = sessionManager()->session();
+
+    auto searchHandlers = session->getData(SEARCH_HANDLERS).value<SearchHandlerDict>();
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+    searchHandler->cancelSearch();
+    searchHandlers.remove(id);
+    session->setData(SEARCH_HANDLERS, QVariant::fromValue(searchHandlers));
+
+    removeActiveSearch(session, id);
+}
+
+void SearchController::categoriesAction()
+{
+    QStringList categories;
+    const QString name = params()["pluginName"].trimmed();
+
+    categories << SearchPluginManager::categoryFullName("all");
+    for (const QString &category : SearchPluginManager::instance()->getPluginCategories(name))
```

Comment:
`for (const QString &category : copyAsConst(SearchPluginManager::instance()->getPluginCategories(name)))`